core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala

```diff
@@ -215,55 +221,265 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
+          rolloverTasks.computeIfAbsent(pr, rolloverRowBuffersTask)
         }
       }
     }
   }
+
+  // Ensure max one background task per table
+  private val rolloverTasks = new ConcurrentHashMap[PartitionedRegion, Future[Unit]]()
+  private val mergeTasks = new ConcurrentHashMap[PartitionedRegion, Future[Unit]]()
+
+  private def minSizeForRollover(pr: PartitionedRegion): Int =
+    math.max(pr.getColumnMaxDeltaRows >>> 3, pr.getColumnMinDeltaRows)
+
+  /**
+   * Check if row buffers are large and have not been touched for a while
+   * then roll it over into the column table
+   */
+  private val rolloverRowBuffersTask = new JFunction[PartitionedRegion, Future[Unit]] {
+    override def apply(pr: PartitionedRegion): Future[Unit] = {
+      val localPrimaries = pr.getDataStore.getAllLocalPrimaryBucketRegions
+      if ((localPrimaries ne null) && localPrimaries.size() > 0) {
+        val doRollover = new Predicate[BucketRegion] {
+          private val minModTime = pr.getCache.cacheTimeMillis() - delayMillis
+
+          override def test(br: BucketRegion): Boolean = {
+            br.getLastModifiedTime <= minModTime && br.getRegionSize >= minSizeForRollover(pr)
+          }
+        }
+        val rolloverBuckets = localPrimaries.asScala.filter(
+          br => doRollover.test(br) && !br.columnBatchFlushLock.isWriteLocked)
+        // enqueue a job to roll over required row buffers into column table
+        // (each bucket will perform a last minute check before rollover inside lock)
+        if (rolloverBuckets.nonEmpty) {
+          implicit val executionContext = Utils.executionContext(pr.getGemFireCache)
+          Future {
+            try {
+              rolloverBuckets.foreach(bucket => withExceptionHandling(
+                bucket.createAndInsertColumnBatch(null, true, doRollover)))
+            } finally {
+              rolloverTasks.remove(pr)
+            }
+          }
+        } else null
+      } else null
+    }
+  }
+
+  /**
+   * Merge multiple column batches that are small in size in a bucket.
+   * These can get created due to a small "tail" in bulk imports (large enough
+   * to exceed minimal size that would have pushed them into row buffers),
+   * or a time-based flush that tolerates small sized column batches due to
+   * [[rolloverRowBuffersTask]] or a forced flush of even smaller size for sample tables.
+   */
+  private def mergeSmallColumnBatches(pr: PartitionedRegion, container: GemFireContainer,
+      metaData: ExternalTableMetaData, smallBatches: mutable.ArrayBuffer[BucketRegion]): Unit = {
+    mergeTasks.computeIfAbsent(pr, new JFunction[PartitionedRegion, Future[Unit]] {
+      override def apply(pr: PartitionedRegion): Future[Unit] = {
+        logInfo(s"Found small batches in ${pr.getName}: ${smallBatches.map(_.getId)}")
+        val cache = pr.getGemFireCache
+        implicit val executionContext = Utils.executionContext(cache)
+        Future(withExceptionHandling({
```

Comment:
That cannot happen because this deliberately uses a "mergeTasks" map with "computeIfAbsent" that will add only one future for a PR at a time. The mergeTasks is cleared for that PR only at the end of Future execution. Same for rolloverTasks.